File: TPS/TP01/Q02-Palindromo-C/palindromo.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <locale.h>
#include <wchar.h>
#include <string.h>

/*
 * void removeLinha(wchar_t str[]) {
 *	size_t len = wcslen(str);
 *
 *	if (len > 0 && str[len - 1] == L'\n') {
 *		str[len - 1] = L'\0';
 *	}
}*
*/

/*
 * bool verificaFim(wchar_t str[], wchar_t chave[]) {
 *	bool fim = false;
 *
 *	//compara se o tamanho da string é igual ao tamanho da chave
 *	if (wcslen(str) == wcslen(chave)) {
 *		fim = true;
 *
 *		//repetição para verificar se a string é igual a chave
 *		for (int i=0; i < wcslen(chave); i++) {
 *			if(str[i] != chave[i]) {
 * 				fim = false;
 *				i = wcslen(chave);
 * 			}
 *		}
 *	}
 *
 *	return fim;
}*
 */
bool ehPalindromo(wchar_t str[]) {
	bool palindromo = true;
	int i = 0,
	    j = (wcslen(str) - 1);

	//repetição para conferir palíndromos
	while (i < j) {
		if (str[i] != str[j]) {
			palindromo = false;
			i = j;
		}
		i++; j--;
	}

	return palindromo;
}

int main() {
	setlocale(LC_ALL, "");
	wchar_t str[100];
	fgetws(str, sizeof(str), stdin);
	str[wcscspn(str, L"\n")] = L'\0';
	

	while (!wcscmp(str, L"FIM")) {
		if (ehPalindromo(str)) {
			wprintf(L"SIM\n");
		}
		else {
			wprintf(L"NAO\n");
		}
		fgetws(str, sizeof(str), stdin);
		str[wcscspn(str, L"\n")] = L'\0';
	}

	return 0;
}

File: TPS/TP01/Q16-Palindromo-Recursivo-JAVA/Palindromo.java
import java.util.Scanner;

public class Palindromo {
	
	public static boolean verificaFim(String str, String chave, int i) {
 		boolean fim = true;		

		if (i >= 0) {
		 	fim = false;

			//compara se o tamanho da string é igual ao tamanho da chave
			if (str.length() == chave.length()) {
				fim = true;

				if (str.charAt(i) != chave.charAt(i)) {
					fim = false;
					i = 0;
				}
				else {
					fim = verificaFim(str, chave, i-1);
				}
			}
			else {
				fim = false;
			}
		} 
	
		return fim;
	}	

	public static boolean ehPalindromo(String str, int i, int j) {
        	boolean palindromo = true;

        	if (i < j) {
                	if (str.charAt(i) != str.charAt(j)) {
                        	palindromo = false;
                        	i = j;
                	}
               		else {
				palindromo = ehPalindromo(str, i+1, j-1);
			}
		}

		return palindromo;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		String str = sc.nextLine().trim();

		while (!verificaFim(str, "FIM", 2)) {
			if (ehPalindromo(str, 0, (str.length()-1))) {
				System.out.println("SIM");
			}
			else {
				System.out.println("NAO");
			}
			str = sc.nextLine().trim();
		}
	}
}
